# SeniorDesign COEN-02
data in '/data/matthewc/Senior_Design/healthcare-dataset-stroke-data.csv'

In [ ]:
import csv
dataset_path = '/data/matthewc/Senior_Design/healthcare-dataset-stroke-data.csv'

In [32]:
with open(dataset_path, 
          newline='') as csvfile:# open thecsv file
    reader = csv.reader(csvfile) #make a reader
    header = reader.__next__()
    datapoint = reader.__next__()
    print(header)


['id', 'gender', 'age', 'hypertension', 'heart_disease', 'ever_married', 'work_type', 'Residence_type', 'avg_glucose_level', 'bmi', 'smoking_status', 'stroke']


In [31]:
for i in range(len(header)):
    print("header[", i,"] =", header[i])

header[ 0 ] = id
header[ 1 ] = gender
header[ 2 ] = age
header[ 3 ] = hypertension
header[ 4 ] = heart_disease
header[ 5 ] = ever_married
header[ 6 ] = work_type
header[ 7 ] = Residence_type
header[ 8 ] = avg_glucose_level
header[ 9 ] = bmi
header[ 10 ] = smoking_status
header[ 11 ] = stroke


In [33]:
print(datapoint)

['9046', 'Male', '67', '0', '1', 'Yes', 'Private', 'Urban', '228.69', '36.6', 'formerly smoked', '1']


Attribute information from kaggle
1) id: unique identifier
2) gender: "Male", "Female" or "Other"
3) age: age of the patient
4) hypertension: 0 if the patient doesn't have hypertension, 1 if the patient has hypertension
5) heart_disease: 0 if the patient doesn't have any heart diseases, 1 if the patient has a heart disease
6) ever_married: "No" or "Yes"
7) work_type: "children", "Govt_jov", "Never_worked", "Private" or "Self-employed"
8) Residence_type: "Rural" or "Urban"
9) avg_glucose_level: average glucose level in blood
10) bmi: body mass index
11) smoking_status: "formerly smoked", "never smoked", "smokes" or "Unknown"*
12) stroke: 1 if the patient had a stroke or 0 if not
*Note: "Unknown" in smoking_status means that the information is unavailable for this patient

Notes:  
 id is not needed for the model, stroke will be the training label.  
 age, hyprtension, heart disease, bmi,and glucose are numerical values.  
 gender - 3 categories  
 ever_married - 2 categories  
 work type - 5 categories  
 residence type - 2 categories  
 smoking status - 4 categories  

One input node is needed for each numrical datapoint --> +5 nodes,  
if one-hot encoding is used then we also need a node per category --> +16 nodes  
21 total input nodes, Can use 1 output node

# Data Processing
- use 1-hot encoding for categorical data
- normalize numeric datapoints

In [2]:
import csv
import torch
dataset_path = '/data/matthewc/Senior_Design/healthcare-dataset-stroke-data.csv'

In [34]:
with open(dataset_path, 
          newline='') as csvfile:# open the csv file
    reader = csv.reader(csvfile) #make a reader
    header = reader.__next__()
    csv_data = reader.__next__()
    print(csv_data)


['9046', 'Male', '67', '0', '1', 'Yes', 'Private', 'Urban', '228.69', '36.6', 'formerly smoked', '1']


21 input nodes - key  
Use floating points for the numeric data:   
0)age = csv_data[2]    
1)hypertension = csv_data[3]  
2)heart_disease = csv_data[4]  
3)glucose_level = csv_data[8]  
4)bmi = csv_data[9] **Has missing values labeled N/A**  

Use one-hot encoding for the categorical data:  
Gender Categories = csv_data[1]  
5)Male  
6)Female  
7)Other  
Ever_Married Categories = csv_data[5]  
8)Yes  
9)No   
Work_type categories = csv_data[6]  
10)Childern  
11)Govt job  
12)Never worked  
13)Private  
14)Self-employed  
Residence type categories = csv_data[7]  
15)Urban  
16)Rural    
Smoking Status Categories = csv_data[10]  
17)Formrly smoked  
18)never smoked  
19)smokes  
20)unknown  

Output = csv_data[11]

In [35]:
new_value = torch.zeros(21)
print(len(new_value))

21


In [36]:
#Numeric data saved as type float
new_value[0]=float(csv_data[2]) #Age
new_value[1]=float(csv_data[3]) #Hypertension
new_value[2]=float(csv_data[4]) #Heart_disease
new_value[3]=float(csv_data[8]) #Glucose_level
if csv_data[9] == 'N/A':
    new_value[4] = 0.0
else: 
    new_value[4]=float(csv_data[9]) #BMI
print(new_value)

tensor([ 67.0000,   0.0000,   1.0000, 228.6900,  36.6000,   0.0000,   0.0000,
          0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,
          0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000,   0.0000])


In [37]:
#Gender Category
if (csv_data[1] == 'Male'):
    new_value[5] = 1.0
    print('Male')
elif (csv_data[1] == 'Female'):
    new_value[6] = 1.0
    print('Female')
elif (csv_data[1] == 'Other'):
    new_value[7] = 1.0
    print('Other')


Male


In [38]:
# Ever Married
if (csv_data[5] == 'Yes'):
    new_value[8] = 1.0
    print('Yes')
elif (csv_data[5] == 'No'):
    new_value[9] = 1.0
    print('No')


Yes


In [39]:
# Work Type
if (csv_data[6] == 'children'):
    new_value[10] = 1.0
    print('children')
elif (csv_data[6] == 'Govt_jov'):
    new_value[11] = 1.0
    print('Govt_jov')
elif (csv_data[6] == 'Never_worked'):
    new_value[12] = 1.0
    print('Never_worked')
elif (csv_data[6] == 'Private'):
    new_value[13] = 1.0
    print('Private')
elif (csv_data[6] == 'Self-employed'):
    new_value[14] = 1.0
    print('Self-employed')


Private


In [40]:
# Residence Type
if (csv_data[7] == 'Urban'):
    new_value[15] = 1.0
    print('Urban')
elif (csv_data[7] == 'Rural'):
    new_value[16] = 1.0
    print('Rural')

Urban


In [41]:
# Smoking Status
if (csv_data[10] == 'formerly smoked'):
    new_value[17] = 1.0
    print('formerly smoked')
elif (csv_data[10] == 'never smoked'):
    new_value[18] = 1.0
    print('never smoked')
elif (csv_data[10] == 'smokes'):
    new_value[19] = 1.0
    print('smokes')
elif (csv_data[10] == 'Unknown'):
    new_value[20] = 1.0
    print('Unknown')

formerly smoked


In [42]:
print(new_value)

tensor([ 67.0000,   0.0000,   1.0000, 228.6900,  36.6000,   1.0000,   0.0000,
          0.0000,   1.0000,   0.0000,   0.0000,   0.0000,   0.0000,   1.0000,
          0.0000,   1.0000,   0.0000,   1.0000,   0.0000,   0.0000,   0.0000])


In [3]:
count = 0
with open(dataset_path, 
          newline='') as csvfile:# open the csv file
    reader = csv.reader(csvfile) #make a reader
    for line in reader:
        csv_data = line
        if(csv_data[0] != 'id'):
            count+=1
print(count)

5110


In [61]:
stack = torch.zeros(count,21)
stack[0,:] = new_value
print(stack.shape)
print(stack[0])

torch.Size([5110, 21])
tensor([ 67.0000,   0.0000,   1.0000, 228.6900,  36.6000,   1.0000,   0.0000,
          0.0000,   1.0000,   0.0000,   0.0000,   0.0000,   0.0000,   1.0000,
          0.0000,   1.0000,   0.0000,   1.0000,   0.0000,   0.0000,   0.0000])


In [23]:
#Loop to create the full dataset
data_stack = torch.zeros(count,21)
label_stack = torch.zeros(count,1)

i = 0
with open(dataset_path, 
          newline='') as csvfile:# open the csv file
    reader = csv.reader(csvfile) #make a reader
    for line in reader:
        csv_data = line
        if(csv_data[0] != 'id'):
            new_value = torch.zeros(21)
            #Numeric data saved as type float
            new_value[0]=float(csv_data[2]) #Age
            new_value[1]=float(csv_data[3]) #Hypertension
            new_value[2]=float(csv_data[4]) #Heart_disease
            new_value[3]=float(csv_data[8]) #Glucose_level
            if csv_data[9] == 'N/A':
                new_value[4] = 0.0
            else: 
                new_value[4]=float(csv_data[9]) #BMI
            # Gender Category
            if (csv_data[1] == 'Male'):
                new_value[5] = 1.0
            elif (csv_data[1] == 'Female'):
                new_value[6] = 1.0
            elif (csv_data[1] == 'Other'):
                new_value[7] = 1.0
            # Ever Married
            if (csv_data[5] == 'Yes'):
                new_value[8] = 1.0
            elif (csv_data[5] == 'No'):
                new_value[9] = 1.0
            # Work Type
            if (csv_data[6] == 'children'):
                new_value[10] = 1.0
            elif (csv_data[6] == 'Govt_jov'):
                new_value[11] = 1.0
            elif (csv_data[6] == 'Never_worked'):
                new_value[12] = 1.0
            elif (csv_data[6] == 'Private'):
                new_value[13] = 1.0
            elif (csv_data[6] == 'Self-employed'):
                new_value[14] = 1.0
                # Residence Type
            if (csv_data[7] == 'Urban'):
                new_value[15] = 1.0
            elif (csv_data[7] == 'Rural'):
                new_value[16] = 1.0
                # Smoking Status
            if (csv_data[10] == 'formerly smoked'):
                new_value[17] = 1.0
            elif (csv_data[10] == 'never smoked'):
                new_value[18] = 1.0
            elif (csv_data[10] == 'smokes'):
                new_value[19] = 1.0
            elif (csv_data[10] == 'Unknown'):
                new_value[20] = 1.0
            data_stack[i,:] = new_value
            label_stack[i,0] = float(csv_data[11])
            i+=1
print("Done")

Done


In [24]:
#torch.save(data_stack, '/data/matthewc/Senior_Design/Full_Dataset.pth')
#torch.save(label_stack, '/data/matthewc/Senior_Design/Labels.pth')

In [45]:
print(data_stack.shape)

torch.Size([5110, 21])


## Missing bmi values

In [30]:
data = torch.load('/data/matthewc/Senior_Design/Full_Dataset.pth', weights_only=True)
count = 0
missing_bmi_count = 0
missing_bmi_idx = []
total_bmi020=0
total_bmi2040=0
total_bmi4060=0
total_bmi6080=0
total_bmi80100=0
count0_20 = 0
count20_40 = 0
count40_60 = 0
count60_80 = 0
count80_100 = 0

total = 0
for value in data:
    count += 1
    total += value[4]
    if (value[4] == 0.0): #Check if the bmi value is missing (set to 0.0)
        missing_bmi_count += 1
        missing_bmi_idx.append(count-1)
    else:    
        if(value[0] >=0 and value[0]<20):
            total_bmi020+=value[4]
            count0_20 += 1
        if(value[0] >=20 and value[0]<40):
            total_bmi2040+=value[4]
            count20_40+=1
        if(value[0] >=40 and value[0]<60):
            total_bmi4060+=value[4]
            count40_60+=1
        if(value[0] >=60 and value[0]<80):
            total_bmi6080+=value[4]
            count60_80+=1
        if(value[0] >=80 and value[0]<100):
            total_bmi80100+=value[4]
            count80_100+=1
        
print("Count = ", count)
print("Missing BMI count = ",missing_bmi_count)
print("Average BMI = ", total/(count-missing_bmi_count))

Count =  5110
Missing BMI count =  201
Average BMI =  tensor(28.8932)


In [34]:
print(missing_bmi_idx)
#These are the datapoints that are missing bmi values

[1, 8, 13, 19, 27, 29, 43, 46, 50, 51, 54, 57, 64, 70, 78, 81, 84, 105, 112, 124, 126, 129, 133, 146, 150, 160, 161, 162, 167, 170, 171, 174, 178, 183, 189, 198, 200, 218, 227, 247, 342, 360, 432, 477, 479, 522, 668, 671, 680, 729, 742, 865, 867, 872, 879, 903, 936, 965, 1102, 1106, 1115, 1183, 1194, 1214, 1235, 1241, 1277, 1293, 1300, 1306, 1324, 1342, 1352, 1427, 1457, 1466, 1468, 1471, 1503, 1528, 1539, 1546, 1596, 1640, 1644, 1646, 1650, 1669, 1670, 1681, 1718, 1719, 1730, 1753, 1756, 1779, 1816, 1836, 1837, 1866, 1894, 1906, 1912, 1981, 1993, 2030, 2103, 2105, 2109, 2192, 2215, 2263, 2285, 2321, 2322, 2339, 2343, 2477, 2494, 2502, 2515, 2529, 2532, 2541, 2582, 2697, 2739, 2752, 2768, 2788, 2816, 2828, 2855, 2867, 2879, 2897, 2914, 2960, 2997, 3007, 3028, 3048, 3059, 3074, 3104, 3111, 3135, 3161, 3162, 3164, 3176, 3197, 3214, 3215, 3216, 3375, 3382, 3425, 3431, 3503, 3562, 3605, 3629, 3681, 3699, 3705, 3726, 3734, 3802, 3808, 3872, 3913, 3940, 3945, 3951, 4046, 4069, 4164, 4202, 42

In [31]:
#Use Age-group mean Imputation to fill in the missing values
group0_20 = []
group20_40 = []
group40_60 = []
group60_80 = []
group80_100 = []

data = torch.load('/data/matthewc/Senior_Design/Full_Dataset.pth', weights_only=True)
for index in missing_bmi_idx:
    if(data[index,0]>=0 and data[index,0]<20):
        group0_20.append(index)
    if(data[index,0]>=20 and data[index,0]<40):
        group20_40.append(index)
    if(data[index,0]>=40 and data[index,0]<60):
        group40_60.append(index)
    if(data[index,0]>=60 and data[index,0]<80):
        group60_80.append(index)
    if(data[index,0]>=80 and data[index,0]<100):
        group80_100.append(index)


In [35]:
print(total_bmi020/count0_20)
print(total_bmi2040/count20_40)
print(total_bmi4060/count40_60)
print(total_bmi6080/count60_80)
print(total_bmi80100/count80_100)

tensor(21.9714)
tensor(29.7921)
tensor(31.6083)
tensor(30.2479)
tensor(28.2813)


In [40]:
data[group0_20,4] = total_bmi020/count0_20
data[group20_40,4] = total_bmi2040/count20_40
data[group40_60,4] = total_bmi4060/count40_60
data[group60_80,4] = total_bmi6080/count60_80
data[group80_100,4] = total_bmi80100/count80_100


In [42]:
for value in data:
    if (value[4] == 0.0): #Check if the bmi value is missing (set to 0.0)
        print("missing bmi value")
print("done")

done


In [43]:
torch.save(data, '/data/matthewc/Senior_Design/Imputed_Dataset.pth')

## Train/Test/Validation data split with SMOTE oversampling

In [1]:
# Import statements
import torch
from imblearn.over_sampling import SMOTE

from torch import nn
from torch.utils.data import DataLoader, Dataset, random_split
import math
import time

In [4]:
class StrokeDataset(torch.utils.data.Dataset):
    def __init__(self, data, labels):
        self.data = data
        self.labels = labels
        
    def __getitem__(self,idx):
        X = self.data[idx,:]
        Y = self.labels[idx,]
        return X, Y

    def __len__(self):
        return self.data.shape[0]   

In [5]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
#Set load the data and labels
data = torch.load('/data/matthewc/Senior_Design/Imputed_Dataset.pth', weights_only=True)
labels = torch.load('/data/matthewc/Senior_Design/Labels.pth', weights_only=True)


In [6]:
sm = SMOTE(random_state=42) #Start tthe SMOTE implementation
resampled_data, resampled_labels =sm.fit_resample(data, labels) #Resample the data with SMOTE
# Convert to pytorch tensors
resampled_data_tensor = torch.from_numpy(resampled_data)
resampled_labels_tensor = torch.from_numpy(resampled_labels)


In [7]:
Full_Dataset = StrokeDataset(data = resampled_data_tensor, labels = resampled_labels_tensor.unsqueeze(1))
#Split the full dataset into seperate training/testing datasets
train_size = Full_Dataset.__len__() *0.7
test_size = Full_Dataset.__len__() *0.15
val_size = Full_Dataset.__len__() *0.15
train_dataset, test_dataset, validation_dataset = random_split(
    Full_Dataset, [int(math.ceil(train_size)), int(math.floor(test_size)), int(math.floor(val_size))],
    generator=torch.Generator().manual_seed(42)
)

In [6]:
print("Full resampled dataset size = ",Full_Dataset.__len__())
print("Training Dataset Size = ", int(math.ceil(train_size)))
print("Testing Dataset Size = ",int(math.floor(test_size)))
print("Validation Dataset Size = ",int(math.floor(val_size)))

Full resampled dataset size =  9722
Training Dataset Size =  6806
Testing Dataset Size =  1458
Validation Dataset Size =  1458


# AI model

In [3]:
class Classifier(nn.Module):
    def __init__(self):
        super(Classifier, self).__init__()
        self.net = nn.Sequential(
            nn.Linear(21, 40), #21 input nodes
            nn.ReLU(),
            nn.Linear(40, 25),
            nn.ReLU(),
            nn.Linear(25, 10),
            nn.ReLU(),
            nn.Linear(10, 2), #2 output nodes for 2 classes
            nn.ReLU()
        )

    def forward(self, x):
        return self.net(x)

In [12]:
#model = Classifier().to(device)
model = torch.load('/data/matthewc/Senior_Design/ReLU_Model.pth', weights_only = False)
#Make dataloders for the datasets
Training_Loader = DataLoader(train_dataset, batch_size = 1, shuffle = True)
Testing_Loader = DataLoader(test_dataset, batch_size = 1, shuffle = True)
Validation_Loader = DataLoader(validation_dataset, batch_size = 1, shuffle = True)

criterion = nn.CrossEntropyLoss()  # Cross Entropy Loss
optimizer = torch.optim.Adam(model.parameters(), lr=0.00005)

In [9]:
epochs = 100
for epoch in range(epochs):
    model.train()
    total_loss = 0
    step = 0
    for ins, labs in Training_Loader: #For inputs/labels
        ins = ins.to(device)
        labs = labs.to(device)
        optimizer.zero_grad()
        output = model(ins)  
        loss = criterion(output, labs.squeeze(dim=0).long())
        loss.backward()
        optimizer.step()
        total_loss += loss.item()
    model.eval()
    correct = 0
    total = 0
    with torch.no_grad():
        for ins, labs in Testing_Loader:
            ins, labs = ins.to(device), labs.float().to(device)
            output = model(ins)
            predicted = output.argmax(dim=1)    
            total += 1
            correct += (predicted == labs).float()
    epoch_loss = total_loss / len(Training_Loader)
    epoch_acc = correct/(total) * 100
    print(f'Epoch {epoch+1}/{epochs}, Loss: {epoch_loss:.4f}, Test Accuracy: {epoch_acc.sum():.2f}%')
print("Done")

Epoch 1/100, Loss: 0.0220, Test Accuracy: 94.99%
Epoch 2/100, Loss: 0.0219, Test Accuracy: 94.44%
Epoch 3/100, Loss: 0.0217, Test Accuracy: 94.58%
Epoch 4/100, Loss: 0.0217, Test Accuracy: 94.65%
Epoch 5/100, Loss: 0.0222, Test Accuracy: 95.06%
Epoch 6/100, Loss: 0.0212, Test Accuracy: 94.99%
Epoch 7/100, Loss: 0.0220, Test Accuracy: 94.99%
Epoch 8/100, Loss: 0.0208, Test Accuracy: 95.13%
Epoch 9/100, Loss: 0.0209, Test Accuracy: 94.72%
Epoch 10/100, Loss: 0.0218, Test Accuracy: 95.40%
Epoch 11/100, Loss: 0.0215, Test Accuracy: 95.13%
Epoch 12/100, Loss: 0.0224, Test Accuracy: 95.13%
Epoch 13/100, Loss: 0.0206, Test Accuracy: 95.27%
Epoch 14/100, Loss: 0.0200, Test Accuracy: 94.44%
Epoch 15/100, Loss: 0.0210, Test Accuracy: 95.34%
Epoch 16/100, Loss: 0.0227, Test Accuracy: 94.65%
Epoch 17/100, Loss: 0.0224, Test Accuracy: 94.92%
Epoch 18/100, Loss: 0.0217, Test Accuracy: 95.06%
Epoch 19/100, Loss: 0.0232, Test Accuracy: 95.27%
Epoch 20/100, Loss: 0.0207, Test Accuracy: 95.06%
Epoch 21/

200 Epochs, ReLU and Sigmoid, 87.31%, Saved in '/data/matthewc/Senior_Design/Baseline_Model.pth'  
200 Epochs, All Sigmoid, 88.48%, Saved in '/data/matthewc/Senior_Design/Sigmoid_Model.pth'  
1000 Epochs, All ReLU, 95.06%, Saved in '/data/matthewc/Senior_Design/ReLU_Model.pth'   
200 Epochs, All tanh, 91.91%  

I started with 200 epoch for each different model and found that the ReLU model was the most promising. 
The 1st and 2nd models started to plateau by 200th epoch, while the 4th model plateaued around 130 epochs in. 
The 3rd model, with only ReLU for the activation function, continued to show steady improvement, so I continued  training it.   
At 1000 epochs it started to plateau aroung 94% accuracy, so I am lowering the learning rate to se if it can improve anymore. 

In [13]:
torch.save(model, '/data/matthewc/Senior_Design/ReLU_Model.pth')

## Validation

In [8]:
model = torch.load('/data/matthewc/Senior_Design/ReLU_Model.pth', weights_only=False)

In [13]:
#Check accuracy on the validation dataset
model.eval()
correct = 0
total = 0
with torch.no_grad():
    for ins, labs in Validation_Loader:
        ins, labs = ins.to(device), labs.float().to(device)
        output = model(ins)
        predicted = output.argmax(dim=1)    
        total += 1
        correct += (predicted == labs).float()
    Validation_Accuracy= (correct/total) * 100
    print(f'Validation dataset accuracy: {Validation_Accuracy.sum():.2f}%')


Validation dataset accuracy: 94.51%


In [14]:
Full_Dataset_Loader = DataLoader(Full_Dataset, batch_size = 1, shuffle = True)
model.eval()
correct = 0
total = 0
with torch.no_grad():
    for ins, labs in Full_Dataset_Loader:
        ins, labs = ins.to(device), labs.float().to(device)
        output = model(ins)
        predicted = output.argmax(dim=1)    
        total += 1
        correct += (predicted == labs).float()
    Validation_Accuracy= (correct/total) * 100
    print(f'Accuracy on full dataset: {Validation_Accuracy.sum():.2f}%')


Accuracy on full dataset: 98.08%


In [15]:
No_SMOTE_Dataset = StrokeDataset(data = data, labels = labels)
No_SMOTE_Loader = DataLoader(No_SMOTE_Dataset, batch_size = 1, shuffle = True)
model.eval()
correct = 0
total = 0
with torch.no_grad():
    for ins, labs in No_SMOTE_Loader:
        ins, labs = ins.to(device), labs.float().to(device)
        output = model(ins)
        predicted = output.argmax(dim=1)    
        total += 1
        correct += (predicted == labs).float()
    Validation_Accuracy= (correct/total) * 100
    print(f'Accuracy on full dataset without oversampling: {Validation_Accuracy.sum():.2f}%')

Accuracy on full dataset without oversampling: 96.85%


In [33]:
model.eval()
true_positive = 0
true_negative = 0
false_positive = 0
false_negative = 0
total = 0

with torch.no_grad():
    for ins, labs in Validation_Loader:
        ins, labs = ins.to(device), labs.float().to(device)
        output = model(ins)
        total += 1
        predicted = output.argmax(dim=1)
        if(predicted == 1 and labs == 1):
            true_positive+=1
        elif(predicted == 0 and labs == 0):
            true_negative+=1        
        elif(predicted == 1 and labs == 0):
            false_positive+=1        
        elif(predicted == 0 and labs == 1):
            false_negative+=1

            
print("Model results on validation dataset")
Accuracy = (true_positive+true_negative)/total *100
print(f'Accuracy : {Accuracy:.2f}%')
Recall = true_positive/(true_positive+false_negative) *100
print(f'Recall : {Recall:.2f}%')
Percision = true_positive/(true_positive+false_positive) *100
print(f'Precision : {Percision:.2f}%')


Model results on validation dataset
Accuracy : 94.51%
Recall : 97.97%
Precision : 91.11%


In [34]:
model.eval()
true_positive = 0
true_negative = 0
false_positive = 0
false_negative = 0
total = 0

with torch.no_grad():
    for ins, labs in No_SMOTE_Loader:
        ins, labs = ins.to(device), labs.float().to(device)
        output = model(ins)
        total += 1
        predicted = output.argmax(dim=1)
        if(predicted == 1 and labs == 1):
            true_positive+=1
        elif(predicted == 0 and labs == 0):
            true_negative+=1        
        elif(predicted == 1 and labs == 0):
            false_positive+=1        
        elif(predicted == 0 and labs == 1):
            false_negative+=1

            
print("Model results on dataset without oversampling")
Accuracy = (true_positive+true_negative)/total *100
print(f'Accuracy : {Accuracy:.2f}%')
Recall = true_positive/(true_positive+false_negative) *100
print(f'Recall : {Recall:.2f}%')
Percision = true_positive/(true_positive+false_positive) *100
print(f'Precision : {Percision:.2f}%')


Model results on dataset without oversampling
Accuracy : 96.85%
Recall : 95.58%
Precision : 61.34%
